docker/Dockerfile

```diff
@@ -0,0 +1,120 @@
+# [Choice] bionic (18.04), focal (20.04)
+ARG VARIANT="focal"
+FROM ubuntu:${VARIANT}
+
+# Restate the variant to use it later on in the llvm and cmake installations
+ARG VARIANT
+
+# Install necessary packages available from standard repos
+RUN apt-get update -qq && export DEBIAN_FRONTEND=noninteractive && \
+    apt-get install -y --no-install-recommends \
+        software-properties-common wget apt-utils file zip \
+        openssh-client gpg-agent socat rsync \
+        make ninja-build git \
+        python3 python3-pip
+
+# Install conan and cmake format
+RUN python3 -m pip install --upgrade pip setuptools && \
+    python3 -m pip install conan cmake_format
+
+# By default, anything you run in Docker is done as superuser.
+# Conan runs some install commands as superuser, and will prepend `sudo` to
+# these commands, unless `CONAN_SYSREQUIRES_SUDO=0` is in your env variables.
+ENV CONAN_SYSREQUIRES_SUDO 0
+# Some packages request that Conan use the system package manager to install
+# a few dependencies. This flag allows Conan to proceed with these installations;
+# leaving this flag undefined can cause some installation failures.
+ENV CONAN_SYSREQUIRES_MODE enabled
```

Comment:
I am actually considering removing all of the `CONAN_*` variables; I think they might fall under the category of "things that aren't needed by all users, and don't belong here". They can easily be added to `.devcontainer/Dockerfile` if necessary.

Reply:
Let's keep here only the bare minimum things that are needed to compile this project. However. I think that conan is needed here. Also, I would like to have consistent behavior between devcontainer and the CI. SO if you remove the conan how the CI will compile this?

Reply:
You raise a good point: We should not release a Dockerfile that cannot build the project. That means we need an extra CI step that tests the Dockerfile, to prove that it can build the project: if it fails this step, the workflow should not be allowed to push the container to GHCI. 

The idea of 'removing Conan' should really be, 'move Conan out of `docker/Dockerfile` and into `.devcontainer/Dockerfile`'. The point of this change would be to allow users the choice of a different package manager, for instance vcpkg (see our sister project https://github.com/aminya/cpp_vcpkg_project). These users would need to extend `docker/Dockerfile` to make it work with vcpkg, but they would be able to do so without polluting their environment variables with Conan-specific variables they don't need, and wasting extra disk space on Conan programs they aren't going to use.

Now that I think about it, I think this is outside the scope of this PR, and that moving Conan-specific code should wait for another PR.

Reply:
The package manager selection could be just another variable in the container. To use the selected package manager, we should just pass the correct `CMAKE_TOOLCHAIN_FILE` generated by the (CMakeToolchain, CMakeDeps) conan and vcpkg. This is the least intrusive way to set up dependencies since I believe that CMakeLists.txt files should be package manager agnostic. Then the `-DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake` can be passed using a 'conan' spesific preset. Well this is outside of the scope of this PR though.